InformationScripting/src/queries/TagQuery.cpp

```diff
@@ -44,94 +48,172 @@ QList<TupleSet> TagQuery::execute(QList<TupleSet> input)
+}, QStringList("TagQuery") + args}, exec_{exec}
+{
+	persistent_ = argument(PERSISTENT_ARGUMENT_NAMES[1]) == "yes";
+}
+
+QList<TupleSet> TagQuery::tags(QList<TupleSet> input)
+{
+	bool addSet = isArgumentSet(ADD_ARGUMENT_NAMES[0]);
+	bool removeSet = isArgumentSet(REMOVE_ARGUMENT_NAMES[0]);
+	Q_ASSERT(!(addSet && removeSet)); // TODO should be user warning
+	if (addSet)
+		return addTags(input);
+	else if (removeSet)
+		return removeTags(input);
+	else
+		return queryTags(input);
+}
 
 QList<TupleSet> TagQuery::queryTags(QList<TupleSet> input)
 {
-	QList<TupleSet> result;
-	QList<TagNode*> foundTags;
-	if (scope() == Scope::Local)
-		foundTags = allTags(target());
-	else if (scope() == Scope::Global)
-		foundTags = allTags();
+	QString tagText = argument(NAME_ARGUMENT_NAMES[0]);
+	Q_ASSERT(tagText.size() > 0); // TODO should be user warning
+	QList<TupleSet> result = input;
+
+	// Querying tags non persistent would just return the tag tuples in the input.
+	if (!persistent_) return input;
+
+	if (scope() == Scope::Local || scope() == Scope::Global)
+	{
+		auto targetNode = scope() == Scope::Local ? target() : nullptr;
+		if (result.empty()) result << TupleSet();
+		insertFoundTags(result[0], matcherFor(tagText), targetNode);
+	}
 	else if (scope() == Scope::Input)
 	{
 		Q_ASSERT(input.size() > 0);
-		TupleSet tupleSet = input.takeFirst();
-		auto astTuples = tupleSet.tuples("ast");
+
+		auto matcher = matcherFor(tagText);
+		auto astTuples = result[0].tuples("ast");
 		for (auto tuple : astTuples)
-		{
-			Model::Node* node = tuple["ast"];
-			if (auto astNode = DCast<Model::CompositeNode>(node))
-			{
-				auto tagExtension = astNode->extension<TagExtension>();
-				if (auto tagNode = tagExtension->tag())
-					foundTags << tagNode;
-			}
-		}
-		result << tupleSet;
+			insertFoundTags(result[0], matcher, tuple["ast"]);
 	}
-
-	for (auto tagNode : foundTags)
-		qDebug() << tagNode->name();
 	return result;
 }
 
 QList<TupleSet> TagQuery::addTags(QList<TupleSet> input)
 {
-	QList<TupleSet> result;
+	QList<TupleSet> result = input;
 	QList<Model::Node*> addTagsTo;
+
+	QString tagText = argument(NAME_ARGUMENT_NAMES[0]);
+	Q_ASSERT(tagText.size() > 0); // TODO should be user warning
+
 	if (scope() == Scope::Local)
 	{
-		// Just add a tag to the target:
+		if (result.empty()) result << TupleSet();
 		addTagsTo << target();
 	}
 	else if (scope() == Scope::Global)
 	{
-		// TODO: does that make sense, to which nodes should we add the tags?
+		// That doesn't make sense, to which nodes should we add the tags?
+		// TODO: warn user
 	}
 	else if (scope() == Scope::Input)
 	{
 		Q_ASSERT(input.size() > 0);
-		TupleSet tupleSet = input.takeFirst();
-		auto astTuples = tupleSet.tuples("ast");
-
-		for (auto tuple : astTuples)
+		for (const auto& tuple : result[0].tuples("ast"))
 			addTagsTo << static_cast<Model::Node*>(tuple["ast"]);
+	}
 
-		result << tupleSet;
+	if (persistent_)
+	{
+		auto treeManager = target()->manager();
```

Comment:
Hmm, a node doesn't necessarily have a manager, but we should probably just assume this here and it simplifies things, and we don't really care about the case where there is no manager. Maybe just put a Q_ASSERT.